In [180]:
#Header list(add it manually)
[ 'UsingIP', 'LongURL', 'ShortURL', 'Symbol@', 'Redirecting//',
           'PrefixSuffix-', 'SubDomains', 'HTTPS', 'DomainRegLen', 'Favicon',
           'NonStdPort', 'HTTPSDomainURL', 'RequestURL', 'AnchorURL',
           'LinksInScriptTags', 'ServerFormHandler', 'InfoEmail', 'AbnormalURL',
           'WebsiteForwarding', 'StatusBarCust', 'DisableRightClick',
           'UsingPopupWindow', 'IframeRedirection', 'AgeofDomain',
           'DNSRecording', 'WebsiteTraffic', 'PageRank', 'GoogleIndex',
           'LinksPointingToPage', 'StatsReport', 'class' ]

['UsingIP',
 'LongURL',
 'ShortURL',
 'Symbol@',
 'Redirecting//',
 'PrefixSuffix-',
 'SubDomains',
 'HTTPS',
 'DomainRegLen',
 'Favicon',
 'NonStdPort',
 'HTTPSDomainURL',
 'RequestURL',
 'AnchorURL',
 'LinksInScriptTags',
 'ServerFormHandler',
 'InfoEmail',
 'AbnormalURL',
 'WebsiteForwarding',
 'StatusBarCust',
 'DisableRightClick',
 'UsingPopupWindow',
 'IframeRedirection',
 'AgeofDomain',
 'DNSRecording',
 'WebsiteTraffic',
 'PageRank',
 'GoogleIndex',
 'LinksPointingToPage',
 'StatsReport',
 'class']

In [181]:
import pandas as pd
import numpy as np

In [182]:
phishing = pd.read_csv(r'phishing.txt')
phishing.head()

-1  1  1.1  1.2  -1.1  -1.2  -1.3  -1.4  -1.5  1.3  ...  1.9  1.10  -1.11  \
0   1  1    1    1     1    -1     0     1    -1    1  ...    1     1     -1   
1   1  0    1    1     1    -1    -1    -1    -1    1  ...    1     1      1   
2   1  0    1    1     1    -1    -1    -1     1    1  ...    1     1     -1   
3   1  0   -1    1     1    -1     1     1    -1    1  ...   -1     1     -1   
4  -1  0   -1    1    -1    -1     1     1    -1    1  ...    1     1      1   

   -1.12  -1.13  -1.14  1.11  1.12  -1.15  -1.16  
0     -1      0     -1     1     1      1     -1  
1     -1      1     -1     1     0     -1     -1  
2     -1      1     -1     1    -1      1     -1  
3     -1      0     -1     1     1      1      1  
4      1      1     -1     1    -1     -1      1  

[5 rows x 31 columns]

In [183]:
phishing.shape

(11054, 31)

In [184]:
phishing.columns = [ 'UsingIP', 'LongURL', 'ShortURL', 'Symbol@', 'Redirecting//',
           'PrefixSuffix-', 'SubDomains', 'HTTPS', 'DomainRegLen', 'Favicon',
           'NonStdPort', 'HTTPSDomainURL', 'RequestURL', 'AnchorURL',
           'LinksInScriptTags', 'ServerFormHandler', 'InfoEmail', 'AbnormalURL',
           'WebsiteForwarding', 'StatusBarCust', 'DisableRightClick',
           'UsingPopupWindow', 'IframeRedirection', 'AgeofDomain',
           'DNSRecording', 'WebsiteTraffic', 'PageRank', 'GoogleIndex',
           'LinksPointingToPage', 'StatsReport', 'class' ]

In [185]:
phishing.head()

UsingIP  LongURL  ShortURL  Symbol@  Redirecting//  PrefixSuffix-  \
0        1        1         1        1              1             -1   
1        1        0         1        1              1             -1   
2        1        0         1        1              1             -1   
3        1        0        -1        1              1             -1   
4       -1        0        -1        1             -1             -1   

   SubDomains  HTTPS  DomainRegLen  Favicon  ...  UsingPopupWindow  \
0           0      1            -1        1  ...                 1   
1          -1     -1            -1        1  ...                 1   
2          -1     -1             1        1  ...                 1   
3           1      1            -1        1  ...                -1   
4           1      1            -1        1  ...                 1   

   IframeRedirection  AgeofDomain  DNSRecording  WebsiteTraffic  PageRank  \
0                  1           -1            -1               0        -1   
1                  1            1            -1               1        -1   
2                  1           -1            -1               1        -1   
3                  1           -1            -1               0        -1   
4                  1            1             1               1        -1   

   GoogleIndex  LinksPointingToPage  StatsReport  class  
0            1                    1            1     -1  
1            1                    0           -1     -1  
2            1                   -1            1     -1  
3            1                    1            1      1  
4            1                   -1           -1      1  

[5 rows x 31 columns]

# Dataset Description

A collection of website URLs for 11000+ websites. Each sample has 30 website parameters and a class label identifying it as a phishing website or not (1 or -1). 

In [186]:
#check for missing data or null values
phishing.isnull().sum()

UsingIP                0
LongURL                0
ShortURL               0
Symbol@                0
Redirecting//          0
PrefixSuffix-          0
SubDomains             0
HTTPS                  0
DomainRegLen           0
Favicon                0
NonStdPort             0
HTTPSDomainURL         0
RequestURL             0
AnchorURL              0
LinksInScriptTags      0
ServerFormHandler      0
InfoEmail              0
AbnormalURL            0
WebsiteForwarding      0
StatusBarCust          0
DisableRightClick      0
UsingPopupWindow       0
IframeRedirection      0
AgeofDomain            0
DNSRecording           0
WebsiteTraffic         0
PageRank               0
GoogleIndex            0
LinksPointingToPage    0
StatsReport            0
class                  0
dtype: int64

In [187]:
#divided into dependent and independent variables
x = phishing.drop('class', axis=1)
y = phishing['class']

In [188]:
#split data into training and testing
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.3, random_state = 101)

In [189]:
#predict using Logistic Regression model
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression(C=100)
logreg.fit(x_train, y_train)
logreg_pred = logreg.predict(x_test)

In [190]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix, classification_report
logreg_Accuracy = accuracy_score(logreg_pred, y_test)
print("Accuracy of Logistic Regression model is : ", logreg_Accuracy)
print(confusion_matrix(logreg_pred, y_test))

Accuracy of Logistic Regression model is :  0.9231233041905336
[[1303  112]
 [ 143 1759]]


In [191]:
print("classification report:")
print(classification_report(logreg_pred, y_test))

classification report:
              precision    recall  f1-score   support

          -1       0.90      0.92      0.91      1415
           1       0.94      0.92      0.93      1902

    accuracy                           0.92      3317
   macro avg       0.92      0.92      0.92      3317
weighted avg       0.92      0.92      0.92      3317



# logistic Regression gives accuracy of 92.3% and when we look into the confusion matrix or classification report, we can see that it is a pretty good model as it classifies high precision, recall and f1-score and also not affected by model bias as it evenly predicts both classes

In [192]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier()
rfc.fit(x_train,y_train)
rfc_pred = rfc.predict(x_test)

In [193]:
rfc_accuracy = accuracy_score(rfc_pred, y_test)
print("Accuracy score of RandomForestClassifier model is : ", rfc_accuracy)
print("confusion matrix: ")
print(confusion_matrix(rfc_pred, y_test))

Accuracy score of RandomForestClassifier model is :  0.9713596623454929
confusion matrix: 
[[1388   37]
 [  58 1834]]


In [194]:
print("classification report:")
print(classification_report(rfc_pred, y_test))

classification report:
              precision    recall  f1-score   support

          -1       0.96      0.97      0.97      1425
           1       0.98      0.97      0.97      1892

    accuracy                           0.97      3317
   macro avg       0.97      0.97      0.97      3317
weighted avg       0.97      0.97      0.97      3317



# Random Forest Classifier gives better accuracy of 96.7% than the logistic regression model and when we look into the confusion matrix or classification report, we can see that it is also a good model as it classifies better high precision, recall and f1-score and also not affected by model bias as it evenly predicts both classes

In [195]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5, metric='minkowski')
knn.fit(x_train, y_train)
knn_pred = knn.predict(x_test)

In [196]:
knn_accuracy = accuracy_score(knn_pred, y_test)
print("Accuracy score of Kneighbors classifier model is : ", rfc_accuracy)
print("confusion matrix")
print(confusion_matrix(knn_pred, y_test))

Accuracy score of Kneighbors classifier model is :  0.9713596623454929
confusion matrix
[[1353   95]
 [  93 1776]]


In [197]:
print("classification report:")
print(classification_report(knn_pred, y_test))

classification report:
              precision    recall  f1-score   support

          -1       0.94      0.93      0.94      1448
           1       0.95      0.95      0.95      1869

    accuracy                           0.94      3317
   macro avg       0.94      0.94      0.94      3317
weighted avg       0.94      0.94      0.94      3317



# K-Neighbors Classifier also gives good accuracy of 96.7% but when we look into the confusion matrix or classification report, we can see that it gives lesser precision, recall and f1-score than Random Forest classifier as it predicts more error values than random forest model though not affected by model bias as it evenly predicts both classes

In [198]:
data = phishing[['PrefixSuffix-', 'AnchorURL']]

In [199]:
train_data, test_data, y_train, y_test = train_test_split(data,y, test_size=0.3, random_state=101)

In [200]:
logreg.fit(train_data,y_train)
data_pred = logreg.predict(test_data)
data_accuracy = accuracy_score(data_pred,y_test)
print("Accuracy score of 2 columns model is : ", data_accuracy)
print("confusion matrix: ")
print(confusion_matrix(data_pred, y_test))

Accuracy score of 2 columns model is :  0.8501658124811576
confusion matrix: 
[[ 959   10]
 [ 487 1861]]


In [201]:
print("classification report:")
print(classification_report(data_pred, y_test))

classification report:
              precision    recall  f1-score   support

          -1       0.66      0.99      0.79       969
           1       0.99      0.79      0.88      2348

    accuracy                           0.85      3317
   macro avg       0.83      0.89      0.84      3317
weighted avg       0.90      0.85      0.86      3317



# When using just two independent variables to predict the class it is seen that it also gives good accuracy of 85.0% and high precision,recall and f1-score, though not as good as using multiple independent variables as previously seen.

# Thuswe have chosen Random Forst Classifier as the final model